src/data_readers/data_reader.cpp

```diff
@@ -724,6 +725,7 @@ void generic_data_reader::instantiate_data_store(const std::vector<int>& local_l
       m_data_store->build_preloaded_owner_map(local_list_sizes);
     }
     preload_data_store();
+    m_data_store->set_is_preloaded();
```

Comment:
If something must be called in every derived class, the Template Method might be a useful implementation strategy.

Reply:
In particular, if you have an algorithm in which there is `common_behavior()` and `custom_derived_behavior()`, you can implement this in this way:
```
class base
{
public:
  void my_algorithm()
  {
    common_behavior();
    custom_derived_behavior();
  }

private:
  void common_behavior() { /* whatever */ }
  virtual void custom_derived_behavior() = 0;
};// class base

class derived : public base
{
  void custom_derived_behavior() override { /* whatever */ }
};// class derived
```
This has many advantages, not least of which is that the contract of the function is enforced in the base class and all the derived class has to do to "get it right" is implement the portion of the behavior that changes. Basically, when someone implements a derived class, they don't have to worry about going through a checklist of assumptions because they're already enforced. This eases both new development and maintenance.

Reply:
> It makes more sense to me if we call `set_is_preloaded` inside of `preload_data_store`, even if it requires repeating that function call for every derived class. 
set_is_preloaded was formerly called set_preload. I changed the name to make it (what I think) is more informative. Anyway, set_preload *was* formerly called in the derived data_reader preload_data_store classes. But since preload_data_store is only called in generic_data_reader, it made sense (again, to me) to put the call there.

If there is a use case where preload_data_store() might be called anyplace other than generice_data_reader::instantiate_data_store(), then I agree with you.

Reply:
The issue is more proper separation and encapsulation of concerns rather than who calls `preload_data_store`. To me, anyway, it seems that setting the preloaded flag is an integral part of preloading the data store. But as implied by the approval, I'm not going to make a fuss about this if you disagree.

Reply:
I will make these changes in a follow on PR.